src/history/database.rs

```diff
@@ -0,0 +1,499 @@
+        debug!("opening sqlite database at {:?}", path);
+
+        let create = !path.exists();
+        if create {
+            if let Some(dir) = path.parent() {
+                std::fs::create_dir_all(dir)?;
+            }
+        }
+
+        let opts = SqliteConnectOptions::from_str(path.as_os_str().to_str().unwrap())?
+            .journal_mode(SqliteJournalMode::Wal)
+            .create_if_missing(true);
+
+        let pool = SqlitePoolOptions::new().connect_with(opts).await?;
+
+        Self::setup_db(&pool).await?;
+        let pid = std::process::id().into();
+        Ok(Self { pool, pid })
+    }
+
+    async fn setup_db(pool: &SqlitePool) -> Result<(), sqlx::Error> {
+        debug!("running sqlite database setup");
+
+        // sqlx::migrate!("./migrations").run(pool).await?;
+
+        //TODO: add run_count
+        //TODO: maybe change command to command_line and then
+        // add command with only the command and parameters as
+        // a separate column in order to do interesting queries
+        let history_table = r#"
+        CREATE TABLE IF NOT EXISTS history_items (
+            history_id   INTEGER PRIMARY KEY NOT NULL,
+            timestamp    INTEGER NOT NULL,
+            duration     INTEGER NOT NULL,
+            exit_status  INTEGER NOT NULL,
+            command      TEXT NOT NULL,
+            cwd          TEXT NOT NULL,
+            session_id   INTEGER NOT NULL,
+
+            UNIQUE(timestamp, cwd, command)
+        );
+
+        CREATE INDEX IF NOT EXISTS idx_history_timestamp on history_items(timestamp);
+        CREATE INDEX IF NOT EXISTS idx_history_command on history_items(command);"#;
+
+        let performance_table = r#"
+        CREATE TABLE IF NOT EXISTS performance_items (
+            perf_id     INTEGER NOT NULL PRIMARY KEY,
+            metrics     FLOAT NOT NULL,
+            history_id  INTEGER NOT NULL
+            REFERENCES history_items(history_id) ON DELETE CASCADE ON UPDATE CASCADE
+          );
+        "#;
+
+        let mut conn = pool.acquire().await?;
+        sqlx::query(history_table).execute(&mut conn).await?;
+        sqlx::query(performance_table).execute(&mut conn).await?;
+
+        Ok(())
+    }
+
+    async fn save_raw(
+        tx: &mut sqlx::Transaction<'_, sqlx::Sqlite>,
+        h: &HistoryItem,
+    ) -> Result<(), sqlx::Error> {
+        // We don't need the history_id here because it's an auto number field
+        // so it should be ever increasing
+        sqlx::query(
+            "insert or ignore into history_items(timestamp, duration, exit_status, command, cwd, session_id)
+                values(?1, ?2, ?3, ?4, ?5, ?6)",
+        )
+        .bind(h.timestamp.timestamp_nanos())
+        .bind(h.duration)
+        .bind(h.exit_status)
+        .bind(h.command.as_str())
+        .bind(h.cwd.as_str())
+        .bind(h.session_id)
+        .execute(tx)
+        .await?;
+
+        Ok(())
+    }
+
+    fn convert_time(h: &HistoryItem) {
```

Comment:
This function isn't used yet but I mocked it up so that it could be used to translate from timestamp_nanos to print out human readable times. This is how save stores datetime.

```rust
    async fn save_raw(
        tx: &mut sqlx::Transaction<'_, sqlx::Sqlite>,
        h: &HistoryItem,
    ) -> Result<(), sqlx::Error> {
        // We don't need the history_id here because it's an auto number field
        // so it should be ever increasing
        sqlx::query(
            "insert or ignore into history_items(timestamp, duration, exit_status, command, cwd, session_id)
                values(?1, ?2, ?3, ?4, ?5, ?6)",
        )
        .bind(h.timestamp.timestamp_nanos())
        .bind(h.duration)
        .bind(h.exit_status)
        .bind(h.command.as_str())
        .bind(h.cwd.as_str())
        .bind(h.session_id)
        .execute(tx)
        .await?;

        Ok(())
    }
```